haystack/preview/components/reader.py

```diff
@@ -0,0 +1,151 @@
+        flat_probabilities = masked_probabilities.flatten(-2, -1)  # necessary for topk
+        candidates = torch.topk(flat_probabilities, top_k)
+        start_candidates = candidates.indices // max_seq_length
+        end_candidates = candidates.indices % max_seq_length
+        start_candidates = start_candidates.cpu()
+        end_candidates = end_candidates.cpu()
+
+        start_candidates = [
+            [encoding.token_to_chars(start)[0] if start != 0 else 0 for start in candidates]
+            for candidates, encoding in zip(start_candidates, encodings)
+        ]
+        end_candidates = [
+            [encoding.token_to_chars(end)[1] if end != 0 else 0 for end in candidates]
+            for candidates, encoding in zip(end_candidates, encodings)
+        ]
+        probabilities = candidates.values.cpu().tolist()
+
+        return start_candidates, end_candidates, probabilities
+
+    def _unflatten(
+        self,
+        start: List[List[int]],
+        end: List[List[int]],
+        probabilities: List[List[float]],
+        flattened_documents: List[str],
+        documents: List[List[Document]],
+        top_k: int,
+    ) -> List[List[Answer]]:
+        answers = [
+            [document[start:end] for start, end in zip(start_candidates_, end_candidates_)]
+            for document, start_candidates_, end_candidates_ in zip(flattened_documents, start, end)
+        ]
+
+        i = 0
+        nested_answers = []
+        for documents_ in documents:
+            current_answers = []
+            j = i + len(documents_)
+            for answers_, probabilities_ in zip(answers[i:j], probabilities[i:j]):
+                for answer, probability in zip(answers_, probabilities_):
+                    current_answers.append(Answer(answer=answer, probability=probability))
+            i = j
+            current_answers = sorted(current_answers, key=lambda answer: answer.probability, reverse=True)[:top_k]
+            nested_answers.append(current_answers)
+
+        return nested_answers
+
+    @component.output_types(answers=List[List[Answer]])
+    def run(
+        self,
+        queries: List[str],
+        documents: List[List[Document]],
+        top_k: Optional[int] = None,
+        max_seq_length: Optional[int] = None,
+    ):
+        top_k = top_k or self.top_k
+        max_seq_length = max_seq_length or self.max_seq_length
+
+        flattened_queries, flattened_documents = self._flatten(queries, documents)
+        input_ids, attention_mask, sequence_ids, encodings = self._preprocess(
+            flattened_queries, flattened_documents, max_seq_length
+        )
+
+        output = self.model(input_ids=input_ids, attention_mask=attention_mask)
+
+        start, end, probabilities = self._postprocess(
+            output, sequence_ids, attention_mask, top_k, encodings, max_seq_length
+        )
+
+        answers = self._unflatten(start, end, probabilities, flattened_documents, documents, top_k)
```

Comment:
I am not sure how nice all of these function calls as all of these functions aren't reusable and also don't completely hide away the complexity. However, I still did it because otherwise I would have a very long method,